### ISSUE TITLE
counter example from python?

### ISSUE THREAD

**Issue by weaversa** [GaloisInc/saw-script#1663]
How can I get a counter-example from a failed proof using the python client? Right now I just get:

```
error: Proof failed.
```

**Comment by weaversa** [comment]
Some more info, when using `view(LogResults(verbose_failure=False))` I get the simple error message. When using `view(LogResults(verbose_failure=True))` I get: 

```
Traceback (most recent call last):
  File "popcount.py", line 46, in test_popcountProof
    popcount_broken1_result = llvm_verify(llvm_module, 'popcount_broken1', popcountContract())
  File "/usr/local/share/saw-script/saw-remote-api/python/saw_client/__init__.py", line 616, in llvm_verify
    view.on_failure(result)
  File "/usr/local/share/saw-script/saw-remote-api/python/saw_client/__init__.py", line 334, in on_failure
    print(self.format_failure(result), file=self.file)
  File "/usr/local/share/saw-script/saw-remote-api/python/saw_client/__init__.py", line 309, in format_failure
    '\t\t' + line for line in failure.exception.stdout.split('\n')
AttributeError: 'NoneType' object has no attribute 'split'
```

**Comment by weaversa** [comment]
Here's an example - `saw` works as expected but `saw-remote-api` gives an error about `split`.

[rotatel.tar.gz](https://github.com/GaloisInc/saw-script/files/8694146/rotatel.tar.gz)

```
$ saw rotatel.saw

[01:39:40.169] Loading file "rotatel.saw"
[01:39:40.253] Verifying rotatel ...
[01:39:40.268] Simulating rotatel ...
[01:39:40.273] Checking proof obligations rotatel ...
[01:39:40.348] Subgoal failed: rotatel safety assertion:
internal: error: in rotatel
Undefined behavior encountered
Details:
  Poison value created
  The second operand of `shl` was equal to or greater than the number of bits in the first operand

[01:39:40.349] SolverStats {solverStatsSolvers = fromList ["SBV->Z3"], solverStatsGoalSize = 52}
[01:39:40.349] ----------Counterexample----------
[01:39:40.349]   a: 2147483648
[01:39:40.349] ----------------------------------
[01:39:40.349] Stack trace:
"llvm_verify" (rotatel.saw:10:19-10:30):
Proof failed.
```

```
$ python3 rotatel.py
E
======================================================================
ERROR: test_rotatelProof (__main__.rotatelTestProof)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "rotatel.py", line 32, in test_rotatelProof
    rotatel_result = llvm_verify(llvm_module, 'rotatel', rotatelContract())
  File "/usr/local/share/saw-script/saw-remote-api/python/saw_client/__init__.py", line 616, in llvm_verify
    view.on_failure(result)
  File "/usr/local/share/saw-script/saw-remote-api/python/saw_client/__init__.py", line 334, in on_failure
    print(self.format_failure(result), file=self.file)
  File "/usr/local/share/saw-script/saw-remote-api/python/saw_client/__init__.py", line 309, in format_failure
    '\t\t' + line for line in failure.exception.stdout.split('\n')
AttributeError: 'NoneType' object has no attribute 'split'

----------------------------------------------------------------------
Ran 1 test in 0.841s

FAILED (errors=1)
🛑  The goal failed to verify.
```

**Comment by weaversa** [comment]
When the dust settles, I'm still interested in getting the counter example (ex, `a: 2147483648`) into a Python variable so I can reason with it.

**Comment by weaversa** [comment]
I sorted part of this - I was running `saw-remote-api` with the `--read-only` flag. So, now we know counter examples are not produced when in --read-only mode. Would it be possible to produce counter examples with this flag set?

**Comment by RyanGlScott** [comment]
Good find. I think there are multiple issues at play here:

1. What should be the amount of counterexample information that SAW Python error messages show by default? That is the subject of #1551.
2. Should `--read-only` suppress counterexamples? Probably not, although there was presumably a reason why `--read-only` was designed to behave this way. I'd have to dig a bit deeper on the `argo` side to figure out why this is the case.
3. The SAW Python client really shouldn't produce an error of `AttributeError: 'NoneType' object has no attribute 'split'`, as that indicates that something has gone awry internally. The culprit are these lines:

   https://github.com/GaloisInc/saw-script/blob/143c48f1e2a9a15d97a24edc32a66eb8ff5ff7f2/saw-remote-api/python/saw_client/__init__.py#L307-L309

   Ultimately, the `exception` is parsed from the JSON representation of [this data type from `argo`](https://github.com/GaloisInc/argo/blob/904fb34872fcef462030fe38978842aa5a9db903/argo/src/Argo.hs#L549-L568). In particular, `stdout` corresponds to `errorStdOut :: Maybe String`. But if `errorStdOut` is `Nothing`, then `stdout` will be `None` on the Python end, which would make invoking `split` blow up.

   In short: we should check that `stdout` is defined before calling `split` here.
